cylc/uiserver/data_mgr.py

```diff
@@ -19,39 +19,133 @@
 import asyncio
 import logging
 
-from cylc.flow.exceptions import ClientError, ClientTimeout
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+from time import sleep
+
 from cylc.flow.network.server import PB_METHOD_MAP
 from cylc.flow.network.scan import MSG_TIMEOUT
+from cylc.flow.network.subscriber import WorkflowSubscriber, process_delta_msg
 from cylc.flow.ws_data_mgr import (
-    EDGES, FAMILIES, FAMILY_PROXIES, JOBS, TASKS, TASK_PROXIES, WORKFLOW
+    EDGES, FAMILIES, FAMILY_PROXIES, JOBS, TASKS, TASK_PROXIES, WORKFLOW,
+    DELTAS_MAP, apply_delta, generate_checksum
 )
+from .workflows_mgr import workflow_request
 
 logger = logging.getLogger(__name__)
 
 
-async def get_workflow_data(w_id, client, method):
-    """Call WS endpoint for entire workflow protobuf message."""
-    # Use already established client
-    try:
-        pb_msg = await client.async_request(method)
-    except ClientTimeout as exc:
-        logger.exception(exc)
-        return (w_id, MSG_TIMEOUT)
-    except ClientError as exc:
-        logger.exception(exc)
-        return (w_id, None)
-    else:
-        ws_data = PB_METHOD_MAP[method]()
-        ws_data.ParseFromString(pb_msg)
-    return (w_id, ws_data)
-
-
 class DataManager:
     """Manage the local data-store acquisition/updates from all workflows."""
 
-    def __init__(self, ws_mgr):
-        self.ws_mgr = ws_mgr
+    def __init__(self, workfloworkflows_mgr):
+        self.workflows_mgr = workfloworkflows_mgr
         self.data = {}
+        self.w_subs = {}
+        self.topics = {topic.encode('utf-8') for topic in DELTAS_MAP}
+        self.loop = None
+
+    async def sync_workflows(self):
+        """Run data store sync with workflow services.
+
+        Subscriptions and sync management is instantiated and run in
+        a separate thread for each workflow. This is to avoid the sync loop
+        blocking the main loop.
+        """
+        self.loop = asyncio.get_running_loop()
+        # Might be options other than threads to achieve
+        # non-blocking subscriptions, but this works.
```

Comment:
Looks really strange, with a ThreadPool within an async function. It wouldn't work if we used the same ioloop instance?

Reply:
How do you run a loop and carry on in parent function otherwise?

Reply:
https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor

Reply:
Oh yeah. Interesting. Ioloop's run_in_executor appears to use a threadpool too

https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor

Reply:
Yeah, and not only is waiting for a publish blocking, but it’s in an infinite loop so when it receives and processes one publish it waits for the next publish... For each workflow...

The other thing; I thought about using asyncio.gather(...), however I don’t think you can dynamically add and remove subscriptions to the coros iterable argument (and you’d still need a different executor)

Reply:
Well, it works as-is, and asyncio uses threading when necessary too. So I think the TODO marker is more than enough for now :-) Let's mark this as resolved :tada:

Reply:
(plus, I use `jupyterhub` + UI Server every day, for several hours, if there's any performance bottleneck, bugs, etc, they will eventually appear and be easy to investigate - our code base is really quite small in ui server for now)

Reply:
I ran this with 32 workflows on my VM running on two cores of my Surface Pro lol! (the workflows slowed it down)

Reply:
I ran this with 32 workflows on my VM running on two cores of my Surface Pro lol! (the workflows slowed it down)